libevm/VMOpt.cpp

```diff
@@ -0,0 +1,154 @@
+
+	cpp-ethereum is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+
+#include "VM.h"
+#include "VMConfig.h"
+#include <libethereum/ExtVM.h>
+using namespace std;
+using namespace dev;
+using namespace dev::eth;
+
+void VM::reportStackUse()
+{
+	static intptr_t p = 0;
+	intptr_t q = intptr_t(&q);
+	if (p)
+		cerr << "STACK: " << p << " - " << q << " = " << (p - q) << endl;
+	p = q;
+}
+
+std::array<InstructionMetric, 256> VM::c_metrics;
+void VM::initMetrics()
+{
+	static bool done=false;
+	if (!done)
+	{
+		for (unsigned i = 0; i < 256; ++i)
+		{
+			InstructionInfo inst = instructionInfo((Instruction)i);
+			c_metrics[i].gasPriceTier = inst.gasPriceTier;
+			c_metrics[i].args = inst.args;
+			c_metrics[i].ret = inst.ret;
+		}
+	}
+	done = true;
+}
+
+//
+// init interpreter on entry
+//
+
+void VM::initEntry()
+{
+	m_bounce = &VM::interpretCases;	
+	m_code_vector = m_ext->code;
+	m_code = m_code_vector.data();
+	
+	initMetrics();	
+	
+	optimize();
+}
+
+int VM::pool_constant(const u256& _con)
+{
+	int i = 0, n = m_pool_vector.size();
+	while (i < n)
+	{
+		if (m_pool_vector[i] == _con)
+			return i;
+	}
+	if (i <= 255 )
+	{
+		m_pool_vector.push_back(_con);
+		return i;
+	}
+	return -1;
+}
+
+void VM::optimize()
+{	
+	// build a table of jump destinations for use in verifyJumpDest
+	for (size_t i = 0; i < m_code_vector.size(); ++i)
+	{
+		byte inst = m_code[i];
+
+		if (inst == (byte)Instruction::JUMPDEST)
+		{
+			m_jumpDests.push_back(i);
+		}
+		else if ((byte)Instruction::PUSH1 <= inst && inst <= (byte)Instruction::PUSH32)
+		{
+			byte n = inst - (byte)Instruction::PUSH1 + 1;
+			i += n;
+		}
+	}
+
+#if defined(EVM_USE_CONSTANT_POOL) || defined (EVM_REPLACE_CONST_JUMP)
+
+	for (size_t i = 0; i < m_code_vector.size(); ++i)
+	{
+		byte inst = m_code[i];
+
+		if ((byte)Instruction::PUSH1 <= inst && inst <= (byte)Instruction::PUSH32)
+		{
+			byte n = inst - (byte)Instruction::PUSH1 + 1;
+
+			// decode pushed bytes to integral value
+			u256 val = m_code[i+1];
+			for (uint64_t j = i+2, m = n; --m; ++j)
+				val = (val << 8) | m_code[j];
+
+#ifdef EVM_USE_CONSTANT_POOL			
+			// add value to constant pool and replace PUSHn with PUSHC if room
+			if (1 < n)
+			{
+				int pool_off = pool_constant(val);
+				if (0 <= pool_off && pool_off < 256)
+				{
+					m_code[i] = (byte)Instruction::PUSHC;
+					m_code[i+1] = (byte)pool_off;
+					m_code[i+2] = n;
+				}
+			}			
+#endif
+
+#ifdef EVM_REPLACE_CONST_JUMP
+			// replace JUMP or JUMPI to constant location with JUMPV or JUMPVI
+			size_t ii = i + n + 1;
+			byte op = m_code[ii];
+			if (op == (byte)Instruction::JUMP)
+			{
+				if (verifyJumpDest(val))
```

Comment:
I am not 100% sure but in that article he talks about doing jit optimizations in the background, so in their case, it does not hurt too much if these optimizations take way too long, because the code is already executed in the foreground.